Public/Src/Cache/ContentStore/Distributed/Redis/RedisDatabaseAdapter.cs

```diff
@@ -286,16 +293,81 @@ public async Task<BoolResult> ExecuteBatchOperationAsync(Context context, IRedis
                         },
                         cancellationToken);
                     await batch.NotifyConsumersOfSuccess();
+
+                    RedisOperationSucceeded();
                     return BoolResult.Success;
                 }
                 catch (Exception ex)
                 {
                     batch.NotifyConsumersOfFailure(ex);
+                    HandleRedisExceptionAndResetMultiplexerIfNeeded(context, ex);
                     return new BoolResult(ex);
                 }
             }
         }
 
+        private void RedisOperationSucceeded()
+        {
+            // If the operation is successful, then reset the error count to 0.
+            Interlocked.Exchange(ref _connectionErrorCount, 0);
+        }
+
+        /// <summary>
+        /// This method "handles" redis connectivity errors and reset the connection multiplexer if there are no successful operations happening.
+        /// </summary>
+        private void HandleRedisExceptionAndResetMultiplexerIfNeeded(Context context, Exception exception)
+        {
+            if (IsRedisConnectionException(exception))
+            {
+                // Using double-checked locking approach to reset the connection multiplexer only once.
+                if (Interlocked.Increment(ref _connectionErrorCount) >= _redisConnectionErrorLimit)

```

Comment:
>if (Interlocked.Increment(ref _connectionErrorCount) >= _redisConnectionErrorLimit) [](start = 16, length = 83)

I'd use == instead of >=. If there's a high volume of concurrent calls, we could end up having the same issue we had with the ContentLocationDatabase's flush.

Reply:
There is another code path incrementing the counter, so if we test for == here we may miss it. (Say threshold is 5. We might see 3 and then 6 because 4 and 5 were incremented by other code path.) Given the double-checked-locking, I don't see how we can get multiple threads trying to reset simultaneously.

---
In reply to: [347636197](https://github.com/microsoft/BuildXL/pull/1229#discussion_r347636197) [](ancestors = 347636197)

Reply:
I can add a comment here that will explain why there is >=. David gave a perfect explanation here.

---
In reply to: [347641974](https://github.com/microsoft/BuildXL/pull/1229#discussion_r347641974) [](ancestors = 347641974,347636197)